micro_sam/training/peft_sam.py

```diff
@@ -0,0 +1,99 @@
+import math
+from typing import List, Union
+
+import torch.nn as nn
+
+from segment_anything.modeling import Sam
+
+
+class LoRASurgery(nn.Module):
+    """Operates on the attention layers for performing low-rank adaptation.
+
+    (Inspired from: https://github.com/JamesQFreeman/Sam_LoRA/)
+
+    In SAM, it is implemented as:
+    ```python
+    self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
+    B, N, C = x.shape
+    qkv = self.qkv(x).reshape(B, N, 3, self.num_heads, self.head_dim).permute(2, 0, 3, 1, 4)
+    q, k, v = qkv.unbind(0)
+    ```
+    """
+    def __init__(
+        self,
+        rank: int,
+        block: nn.Module,
+    ):
+        super().__init__()
+        self.qkv = block.attn.qkv
+        self.dim = self.qkv.in_features
+
+        self.w_a_linear_q = nn.Linear(self.dim, rank, bias=False)
+        self.w_b_linear_q = nn.Linear(rank, self.dim, bias=False)
+        self.w_a_linear_v = nn.Linear(self.dim, rank, bias=False)
+        self.w_b_linear_v = nn.Linear(rank, self.dim, bias=False)
+
+        self.reset_parameters()
```

Comment:
Add the update to the block here. I.e.
```
blk.attn.qkv = self
```
This way the updates to the encoder are fully encapsulated here, and we have more flexibility and could e.g. later add a PEFT block that also updates the linear layers.

Reply:
Oh yes, this would have it completely independent. Thanks for the suggestion!